### DISCUSSION TITLE
tasmota32 development builds failing

### DISCUSSION THREAD

**Discussion by MaceMoneta** [arendst/Tasmota#14474] [Q&A] [answered]
I just noticed that tasmota32 development builds (e.g.tasmota32-webcam) are failing to build. I git cloned a pristine directory, same. Also tried the zip archive, no change. From the errors, it looks like one or more missing libraries:

Linking .pio/build/tasmota32-webcam/firmware.elf
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0x48): undefined reference to `digitalRead'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0x98): undefined reference to `WiFiClass32::setOutputPower(int)'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0x9c): undefined reference to `WiFiClass32::setSleepMode(int)'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0xa0): undefined reference to `__pinMode'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0xa8): undefined reference to `String::operator[](unsigned int)'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0xbc): undefined reference to `micros'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0x11c): undefined reference to `be_gc_collect'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0x120): undefined reference to `WiFiSTAClass::BSSIDstr()'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0x124): undefined reference to `WiFiGenericClass::channel()'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0x174): undefined reference to `EspClass::restart()'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0x204): undefined reference to `psramFound'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0x214): undefined reference to `EspClass::getSketchSize()'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0x218): undefined reference to `EspClass::getEfuseMac()'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0x2d8): undefined reference to `getCpuFrequencyMhz'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0x2dc): undefined reference to `EspClass::getFlashChipSpeed()'
/btrfs/serenityRelocated/home/mmoneta/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /tmp/ccUSsm1Q.ltrans3.ltrans.o:(.literal+0x2e0): undefined reference to `EspClass::getFlashChipMode()'
...

**Comment by barbudor** [comment, marked as answer]
Just did a clean build in a fresh Gitpod and all good:
```
Linking .pio/build/tasmota32-webcam/firmware.elf
Retrieving maximum program size .pio/build/tasmota32-webcam/firmware.elf
Checking size .pio/build/tasmota32-webcam/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [==        ]  20.0% (used 65640 bytes from 327680 bytes)
Flash: [=======   ]  66.2% (used 1258153 bytes from 1900544 bytes)
Building .pio/build/tasmota32-webcam/firmware.bin
esptool.py v3.2
Merged 25 ELF sections
esp32_create_combined_bin([".pio/build/tasmota32-webcam/firmware.bin"], [".pio/build/tasmota32-webcam/firmware.elf"])
    Offset | File
 -  0x1000 | /home/gitpod/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32/bin/bootloader_dout_80m.bin
 -  0x8000 | /workspace/Tasmota/.pio/build/tasmota32-webcam/partitions.bin
 -  0xe000 | /home/gitpod/.platformio/packages/framework-arduinoespressif32/tools/partitions/boot_app0.bin
 - 0x10000 | /workspace/Tasmota/.pio/build/tasmota32-webcam/firmware.bin
esptool.py v3.2
Wrote 0x1449a0 bytes to file /workspace/Tasmota/.pio/build/tasmota32-webcam/firmware.factory.bin, ready to flash to offset 0x0
bin_map_copy([".pio/build/tasmota32-webcam/firmware.bin"], [".pio/build/tasmota32-webcam/firmware.elf"])
map_gzip([".pio/build/tasmota32-webcam/firmware.bin"], [".pio/build/tasmota32-webcam/firmware.elf"])
====================================== [SUCCESS] Took 90.38 seconds ======================================

Environment       Status    Duration
----------------  --------  ------------
tasmota32-webcam  SUCCESS   00:01:30.383
======================================= 1 succeeded in 00:01:30.383 =======================================

***********************************************************************************************************
We found 179.00MB of unnecessary PlatformIO system data (temporary files, unnecessary packages, etc.).
Use `pio system prune --dry-run` to list them or `pio system prune` to save disk space.
```

Something gets corrupted in your environment. That happens with PlatformIO.
I suggest `rm -Rf ~/.platformio .pio/ .cache/` and run again
If using VSCode, you may need to quit and reload in order for PlatformIO extension to redownload the missing bits.

**Comment by MaceMoneta** [comment]
Yup, that did it. Thanks, first time I ran into that.
